microscope/devices.py

```diff
@@ -409,6 +409,8 @@ def __init__(self, buffer_length=0, **kwargs):
         self._acquiring = False
         # A condition to signal arrival of a new data and unblock grab_next_data
         self._new_data_condition = threading.Condition()
+        # A data processing pipeline: a list of f(data) -> data.
+        self.pipeline = []
```

Comment:
Should we not make this a private attribute? Also, would be nice to also have type annotations for it.

Reply:
Maybe it should be private, but then we'd need methods to allow other devices to add their processing steps.

Reply:
We talked about this at length.
* Most (all?) classes derived from DataDevice should probably add to the pipeline rather than override _process_data.
* Devices that override _process_data must call super()._process_data().
* The docstring on DataDevice._process_data should indicate this clearly.
* We could make pipeline private, but that would require methods to add/remove steps from it.

Reply:
```_process_data``` docstring made more explicit in 1c16359.
```CameraDevice``` made to use ```pipeline``` instead of override in 0361a02.
```_process_data``` is not only defined in one place, on the ```DataDevice``` class.